New test: [iOS macOS] TestIPC.StreamConnectionTest.SendAsyncReplyMessage is timing out

kkinnunen-apple opened the pull request:
#### e6b67655bc4fb2dcbc0e09b6024792603c6e0261
<pre>
New test: [iOS macOS] TestIPC.StreamConnectionTest.SendAsyncReplyMessage is timing out
<a href="https://bugs.webkit.org/show_bug.cgi?id=249780">https://bugs.webkit.org/show_bug.cgi?id=249780</a>
rdar://103640558

Reviewed by Matt Woodrow.

The test would time out mainly because the communication semaphores were not
set up between client and server connection.
If the server would be slow enough, the initial processing wakeup due to
starting message listening would make the test pass.
If the server would be fast enough, the initial processing wakeup would
not process all messages that the test would send, and thus it would
time out. Fix by setting the semaphores, so new messages will wake up
the server work queue.

Secondarily make the test more consistent by running all the server
related code in the server work queue.

The first point needs IPCSemaphore to have copy operations, so implement
these. Add tests for IPCSemaphore, test also the new operations.
These semaphores are typically sent as part of the initialization message
exchange between client and server (CreateXXX, DidCreateXXX messages).
The unit tests run in single process, there is no message exchange
to setup the connection.

The second point needs a fix for StreamConnectionWorkQueue, where
the tasks that would be run during shutdown would crash if the
tasks would assert that the task is run on correct queue. Fix by
not removing the m_processingThread reference until the thread
has exited. Add a test for this.

* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
(IPC::StreamConnectionWorkQueue::stopAndWaitForCompletion):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::copyViaEncoder):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::StreamConnectionTest::localReferenceBarrier):
(TestWebKitAPI::StreamConnectionTest::serverQueue):
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::StreamConnectionTest::StreamConnectionTest): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionWorkQueueTests.cpp: Added.
(TestWebKitAPI::TEST_F):

Canonical link: <a href="https://commits.webkit.org/258913@main">https://commits.webkit.org/258913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e436a82ff504e175225076ebee2e170bb63f8e62

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103334 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36299 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112570 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172774 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107288 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13482 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3360 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95551 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111312 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109109 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10355 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37987 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92175 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25022 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79718 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5846 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26425 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6019 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2955 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12001 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45934 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6126 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7777 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
```diff
@@ -32,6 +32,15 @@
 namespace TestWebKitAPI {
 
+template <typename T>
+std::optional<T> copyViaEncoder(const T& o)
```
Could you just use this to create same-process copies of the Semaphore for the test code, and then not need to add the copy constructor?

It seems fine either way though, to support both in-process copies as well as IPC ones, just adds a bit more code that's only used for testing.

kkinnunen-apple replied:
> Could you just use this to create same-process copies of the Semaphore for the test code, and then not need to add the copy constructor?

While contrary to the goal of bringing order to (my) world, I think your suggestion is unfortunately the better approach.

kkinnunen-apple committed:
New test: [iOS macOS] TestIPC.StreamConnectionTest.SendAsyncReplyMessage is timing out
https://bugs.webkit.org/show_bug.cgi?id=249780
rdar://103640558

Reviewed by Matt Woodrow.

The test would time out mainly because the communication semaphores were not
set up between client and server connection.
If the server would be slow enough, the initial processing wakeup due to
starting message listening would make the test pass.
If the server would be fast enough, the initial processing wakeup would
not process all messages that the test would send, and thus it would
time out. Fix by setting the semaphores, so new messages will wake up
the server work queue.

Secondarily make the test more consistent by running all the server
related code in the server work queue.

The first point needs IPCSemaphore to have copy operations, so implement
these. Add tests for IPCSemaphore, test also the new operations.
These semaphores are typically sent as part of the initialization message
exchange between client and server (CreateXXX, DidCreateXXX messages).
The unit tests run in single process, there is no message exchange
to setup the connection.

The second point needs a fix for StreamConnectionWorkQueue, where
the tasks that would be run during shutdown would crash if the
tasks would assert that the task is run on correct queue. Fix by
not removing the m_processingThread reference until the thread
has exited. Add a test for this.

* Source/WebKit/Platform/IPC/StreamConnectionWorkQueue.cpp:
(IPC::StreamConnectionWorkQueue::stopAndWaitForCompletion):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::copyViaEncoder):
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionTests.cpp:
(TestWebKitAPI::StreamConnectionTest::localReferenceBarrier):
(TestWebKitAPI::StreamConnectionTest::serverQueue):
(TestWebKitAPI::TEST_F):
(TestWebKitAPI::StreamConnectionTest::StreamConnectionTest): Deleted.
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionWorkQueueTests.cpp: Added.
(TestWebKitAPI::TEST_F):

Canonical link: https://commits.webkit.org/258913@main
